### PR TITLE
Rewrite BaseAssertionAnalyzer as an IOperation analyzer

### PR DESCRIPTION
**Pull request by Dreamescaper**
Contributes to #270 .

Code is ready, but I'm still not sure how to measure performance effect of those changes.

The most significant changes are in
[BaseAssertionAnalyzer.cs](https://github.com/nunit/nunit.analyzers/pull/282/files#diff-a1eba04d22bcead6adfb89aa06669611)
[AssertHelper.cs](https://github.com/nunit/nunit.analyzers/pull/282/files#diff-fa698baaf6c63427affd2527b69f0511)
[ConstraintExpression.cs](https://github.com/nunit/nunit.analyzers/pull/282/files#diff-4e1b0aeab5293667008bdd3ebf98aeb3)
[ConstraintExpressionPart.cs](https://github.com/nunit/nunit.analyzers/pull/282/files#diff-6604d3d8e996a11735d4a6f8326eade1)

### CODE REVIEW

**Comment by mikkelbu** [comment]
I don't have more time today, but hopefully I can take a look at it the next couple of days. Regarding measuring the performance effect, then I was thinking to use the method described in https://m0sa.net/posts/2019-08-roslyn-analyzer-runtime-performance/ and the trying to build a project with a lot of tests/asserts, e.g. the NUnit project or another project (if we cannot find one large enough, we can also write some code/T4 template to generate a lot of tests/asserts).

**Comment by Dreamescaper** [comment]
NUnit project
<details>
<summary>Before - 6.5 seconds</summary>

    6.530   99   nunit.analyzers, Version=0.5.1.0, Culture=neutral, PublicKeyToken=null
    0.996   15      NUnit.Analyzers.TestCaseUsage.TestCaseUsageAnalyzer
    0.860   13      NUnit.Analyzers.StringConstraintWrongActualType.StringConstraintWrongActualTypeAnalyzer
    0.662   10      NUnit.Analyzers.TestCaseUsage.TestMethodUsageAnalyzer
    0.565    8      NUnit.Analyzers.EqualToIncompatibleTypes.EqualToIncompatibleTypesAnalyzer
    0.393    6      NUnit.Analyzers.ContainsConstraintWrongActualType.ContainsConstraintWrongActualTypeAnalyzer
    0.381    5      NUnit.Analyzers.SameActualExpectedValue.SameActualExpectedValueAnalyzer
    0.362    5      NUnit.Analyzers.SomeItemsIncompatibleTypes.SomeItemsIncompatibleTypesAnalyzer
    0.352    5      NUnit.Analyzers.IgnoreCaseUsage.IgnoreCaseUsageAnalyzer
    0.326    4      NUnit.Analyzers.NullConstraintUsage.NullConstraintUsageAnalyzer
    0.319    4      NUnit.Analyzers.MissingProperty.MissingPropertyAnalyzer
    0.240    3      NUnit.Analyzers.ClassicModelAssertUsage.ClassicModelAssertUsageAnalyzer
    0.235    3      NUnit.Analyzers.ConstActualValueUsage.ConstActualValueUsageAnalyzer
    0.224    3      NUnit.Analyzers.SameAsIncompatibleTypes.SameAsIncompatibleTypesAnalyzer
    0.208    3      NUnit.Analyzers.TestCaseSourceUsage.TestCaseSourceUsesStringAnalyzer
    0.155    2      NUnit.Analyzers.TestMethodAccessibilityLevel.TestMethodAccessibilityLevelAnalyzer
    0.141    2      NUnit.Analyzers.ParallelizableUsage.ParallelizableUsageAnalyzer
    0.112    1      NUnit.Analyzers.ValueSourceUsage.ValueSourceUsageAnalyzer
    <0.001   <1      NUnit.Analyzers.ConstraintUsage.EqualConstraintUsageAnalyzer
    <0.001   <1      NUnit.Analyzers.ConstraintUsage.StringConstraintUsageAnalyzer
    <0.001   <1      NUnit.Analyzers.NonTestMethodAccessibilityLevel.NonTestMethodAccessibilityLevelAnalyzer
    <0.001   <1      NUnit.Analyzers.ConstraintUsage.SomeItemsConstraintUsageAnalyzer
</details>


<details>
<summary>After - 3.3 seconds</summary>

       3.311   99   nunit.analyzers, Version=0.5.2.0, Culture=neutral, PublicKeyToken=null
       0.951   28      NUnit.Analyzers.TestCaseUsage.TestCaseUsageAnalyzer
       0.617   18      NUnit.Analyzers.TestCaseUsage.TestMethodUsageAnalyzer
       0.338   10      NUnit.Analyzers.TestCaseSourceUsage.TestCaseSourceUsesStringAnalyzer
       0.250    7      NUnit.Analyzers.EqualToIncompatibleTypes.EqualToIncompatibleTypesAnalyzer
       0.213    6      NUnit.Analyzers.TestMethodAccessibilityLevel.TestMethodAccessibilityLevelAnalyzer
       0.129    3      NUnit.Analyzers.NullConstraintUsage.NullConstraintUsageAnalyzer
       0.116    3      NUnit.Analyzers.ValueSourceUsage.ValueSourceUsageAnalyzer
       0.107    3      NUnit.Analyzers.StringConstraintWrongActualType.StringConstraintWrongActualTypeAnalyzer
       0.105    3      NUnit.Analyzers.ClassicModelAssertUsage.ClassicModelAssertUsageAnalyzer
       0.105    3      NUnit.Analyzers.ParallelizableUsage.ParallelizableUsageAnalyzer
       0.062    1      NUnit.Analyzers.IgnoreCaseUsage.IgnoreCaseUsageAnalyzer
       0.061    1      NUnit.Analyzers.SomeItemsIncompatibleTypes.SomeItemsIncompatibleTypesAnalyzer
       0.059    1      NUnit.Analyzers.SameActualExpectedValue.SameActualExpectedValueAnalyzer
       0.057    1      NUnit.Analyzers.ContainsConstraintWrongActualType.ContainsConstraintWrongActualTypeAnalyzer
       0.056    1      NUnit.Analyzers.MissingProperty.MissingPropertyAnalyzer
       0.047    1      NUnit.Analyzers.SameAsIncompatibleTypes.SameAsIncompatibleTypesAnalyzer
       0.038    1      NUnit.Analyzers.ConstActualValueUsage.ConstActualValueUsageAnalyzer
      <0.001   <1      NUnit.Analyzers.ConstraintUsage.SomeItemsConstraintUsageAnalyzer
      <0.001   <1      NUnit.Analyzers.ConstraintUsage.EqualConstraintUsageAnalyzer
      <0.001   <1      NUnit.Analyzers.NonTestMethodAccessibilityLevel.NonTestMethodAccessibilityLevelAnalyzer
      <0.001   <1      NUnit.Analyzers.ConstraintUsage.StringConstraintUsageAnalyzer

</details>



My work project
<details>
<summary>Before - 16.4 seconds</summary>

    16.452   26   nunit.analyzers, Version=0.5.1.0, Culture=neutral, PublicKeyToken=null
    2.049    3      NUnit.Analyzers.StringConstraintWrongActualType.StringConstraintWrongActualTypeAnalyzer
    1.471    2      NUnit.Analyzers.EqualToIncompatibleTypes.EqualToIncompatibleTypesAnalyzer
    1.454    2      NUnit.Analyzers.TestCaseUsage.TestCaseUsageAnalyzer
    1.332    2      NUnit.Analyzers.NullConstraintUsage.NullConstraintUsageAnalyzer
    1.316    2      NUnit.Analyzers.SomeItemsIncompatibleTypes.SomeItemsIncompatibleTypesAnalyzer
    1.165    1      NUnit.Analyzers.ConstActualValueUsage.ConstActualValueUsageAnalyzer
    1.130    1      NUnit.Analyzers.ContainsConstraintWrongActualType.ContainsConstraintWrongActualTypeAnalyzer
    1.073    1      NUnit.Analyzers.MissingProperty.MissingPropertyAnalyzer
    1.014    1      NUnit.Analyzers.IgnoreCaseUsage.IgnoreCaseUsageAnalyzer
    0.997    1      NUnit.Analyzers.TestCaseUsage.TestMethodUsageAnalyzer
    0.972    1      NUnit.Analyzers.SameAsIncompatibleTypes.SameAsIncompatibleTypesAnalyzer
    0.936    1      NUnit.Analyzers.ClassicModelAssertUsage.ClassicModelAssertUsageAnalyzer
    0.930    1      NUnit.Analyzers.SameActualExpectedValue.SameActualExpectedValueAnalyzer
    0.183   <1      NUnit.Analyzers.TestMethodAccessibilityLevel.TestMethodAccessibilityLevelAnalyzer
    0.147   <1      NUnit.Analyzers.TestCaseSourceUsage.TestCaseSourceUsesStringAnalyzer
    0.146   <1      NUnit.Analyzers.ValueSourceUsage.ValueSourceUsageAnalyzer
    0.136   <1      NUnit.Analyzers.ParallelizableUsage.ParallelizableUsageAnalyzer
    <0.001   <1      NUnit.Analyzers.ConstraintUsage.EqualConstraintUsageAnalyzer
    <0.001   <1      NUnit.Analyzers.ConstraintUsage.StringConstraintUsageAnalyzer
    <0.001   <1      NUnit.Analyzers.NonTestMethodAccessibilityLevel.NonTestMethodAccessibilityLevelAnalyzer
    <0.001   <1      NUnit.Analyzers.ConstraintUsage.SomeItemsConstraintUsageAnalyzer

</details>

<details>
<summary>After - 5.3 seconds</summary>

       5.283   10   nunit.analyzers, Version=0.5.2.0, Culture=neutral, PublicKeyToken=null
       1.653    3      NUnit.Analyzers.TestCaseUsage.TestCaseUsageAnalyzer
       0.797    1      NUnit.Analyzers.TestCaseUsage.TestMethodUsageAnalyzer
       0.286   <1      NUnit.Analyzers.EqualToIncompatibleTypes.EqualToIncompatibleTypesAnalyzer
       0.283   <1      NUnit.Analyzers.NullConstraintUsage.NullConstraintUsageAnalyzer
       0.255   <1      NUnit.Analyzers.StringConstraintWrongActualType.StringConstraintWrongActualTypeAnalyzer
       0.229   <1      NUnit.Analyzers.TestCaseSourceUsage.TestCaseSourceUsesStringAnalyzer
       0.227   <1      NUnit.Analyzers.IgnoreCaseUsage.IgnoreCaseUsageAnalyzer
       0.216   <1      NUnit.Analyzers.ContainsConstraintWrongActualType.ContainsConstraintWrongActualTypeAnalyzer
       0.215   <1      NUnit.Analyzers.ParallelizableUsage.ParallelizableUsageAnalyzer
       0.208   <1      NUnit.Analyzers.SomeItemsIncompatibleTypes.SomeItemsIncompatibleTypesAnalyzer
       0.202   <1      NUnit.Analyzers.SameActualExpectedValue.SameActualExpectedValueAnalyzer
       0.168   <1      NUnit.Analyzers.TestMethodAccessibilityLevel.TestMethodAccessibilityLevelAnalyzer
       0.152   <1      NUnit.Analyzers.SameAsIncompatibleTypes.SameAsIncompatibleTypesAnalyzer
       0.146   <1      NUnit.Analyzers.MissingProperty.MissingPropertyAnalyzer
       0.116   <1      NUnit.Analyzers.ValueSourceUsage.ValueSourceUsageAnalyzer
       0.070   <1      NUnit.Analyzers.ConstActualValueUsage.ConstActualValueUsageAnalyzer
       0.060   <1      NUnit.Analyzers.ClassicModelAssertUsage.ClassicModelAssertUsageAnalyzer
      <0.001   <1      NUnit.Analyzers.NonTestMethodAccessibilityLevel.NonTestMethodAccessibilityLevelAnalyzer
      <0.001   <1      NUnit.Analyzers.ConstraintUsage.EqualConstraintUsageAnalyzer
      <0.001   <1      NUnit.Analyzers.ConstraintUsage.SomeItemsConstraintUsageAnalyzer
      <0.001   <1      NUnit.Analyzers.ConstraintUsage.StringConstraintUsageAnalyzer

</details>

Up to 3 times improvement on total analysis time. Sounds convincing :)

File: src/nunit.analyzers/Operations/ConstraintExpression.cs [hunk 1]
```diff
@@ -1,63 +1,65 @@
...
 using NUnit.Analyzers.Extensions;
 
-namespace NUnit.Analyzers.Syntax
+namespace NUnit.Analyzers.Operations
 {
     /// <summary>
     /// Represents assert constraint expression, e.g. 'Is.EqualTo(expected)', 'Is.Not.Null & Is.Not.Empty'
     /// </summary>
     internal class ConstraintExpression
     {
```
Comment (manfred-brands) [inline src/nunit.analyzers/Operations/ConstraintExpression.cs:14]: Suggest type rename to ConstraintOperation (similar as you did for test class)

File: src/nunit.analyzers/Operations/ConstraintExpressionPart.cs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+using NUnit.Analyzers.Extensions;
+
+namespace NUnit.Analyzers.Operations
+{
+    /// <summary>
+    /// Represents part of <see cref="ConstraintExpression"/>, which is combined with others with 
+    /// binary operators ('&', '|')  or methods ('And', 'Or', 'With').
+    /// </summary>
+    internal class ConstraintExpressionPart
+    {
```
Comment (manfred-brands) [inline src/nunit.analyzers/Operations/ConstraintExpressionPart.cs:16]: Rename class to ConstraintOperationPart matching renamed tests.

File: src/nunit.analyzers/ConstraintUsage/SomeItemsConstraintUsageAnalyzer.cs [hunk 3]
```diff
@@ -27,20 +27,21 @@ public class SomeItemsConstraintUsageAnalyzer : BaseConditionConstraintAnalyzer
...
+                if ((argumentCount == 1 && this.IsCollectionContains(symbol))
+                    || (argumentCount == 2 && this.IsLinqContains(symbol)))
                 {
                     var suggestedConstraint = negated ? DoesNotContain : DoesContain;
                     return (descriptor, suggestedConstraint);
                 }
             }
+
             return (null, null);
         }
```
Comment (manfred-brands) [inline src/nunit.analyzers/ConstraintUsage/SomeItemsConstraintUsageAnalyzer.cs:47]: In 2 methods below change parameter type to IMethodSymbol and drop ' is ' check.

File: src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs [hunk 4]
```diff
@@ -24,90 +22,74 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
...
-                    string.Equals(memberAccessExpression.Expression.ToString(), "string", StringComparison.OrdinalIgnoreCase) &&
-                    string.Equals(memberAccessExpression.Name.Identifier.Text, "Empty", StringComparison.Ordinal);
+                return operation is IFieldReferenceOperation propertyReference
+                    && propertyReference.Field.Name == nameof(string.Empty)
+                    && propertyReference.Field.ContainingType.SpecialType == SpecialType.System_String;
             }
 
-            void Report(ExpressionSyntax expression)
+            void Report(SyntaxNode expression)
             {
```
Comment (manfred-brands) [inline src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:49]: Change parameter to IOperation. This then moved the .Syntax call from the callers to the method.

File: src/nunit.analyzers/Operations/ConstraintExpression.cs (hunk 1 above)
Comment (Dreamescaper) [reply src/nunit.analyzers/Operations/ConstraintExpression.cs:14]: I thought about that, but `Expression` is not a Roslyn term here, but NUnit.
See https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/ConstraintExpression.cs

Comment (manfred-brands) [reply src/nunit.analyzers/Operations/ConstraintExpression.cs:14]: OK, then you should rename the unit test back to ConstraintExpressionTests.

Comment (Dreamescaper) [reply src/nunit.analyzers/Operations/ConstraintExpression.cs:14]: Yes, agree

File: src/nunit.analyzers/SameAsIncompatibleTypes/SameAsIncompatibleTypesAnalyzer.cs [hunk 5]
```diff
@@ -23,53 +23,46 @@ public class SameAsIncompatibleTypesAnalyzer : BaseAssertionAnalyzer
...
-            var cancellationToken = context.CancellationToken;
-            var semanticModel = context.SemanticModel;
 
-            if (!AssertHelper.TryGetActualAndConstraintExpressions(assertExpression, semanticModel,
-                out var actualExpression, out var constraintExpression))
+        protected override void AnalyzeAssertInvocation(OperationAnalysisContext context, IInvocationOperation assertOperation)
+        {
+            if (!AssertHelper.TryGetActualAndConstraintOperations(assertOperation,
+                out var actualOperation, out var constraintOperation))
             {
```
Comment (manfred-brands) [inline src/nunit.analyzers/SameAsIncompatibleTypes/SameAsIncompatibleTypesAnalyzer.cs:31]: Can you undo the rename and rename it back to constraintExpression?

File: src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs [hunk 6]
```diff
@@ -24,90 +22,74 @@ public class ConstActualValueUsageAnalyzer : BaseAssertionAnalyzer
...
-            if (expectedExpression != null && !IsConstant(expectedExpression) && !IsStringEmpty(expectedExpression))
+            if (expectedExpression != null && !expectedExpression.ConstantValue.HasValue && !IsStringEmpty(expectedExpression))
             {
-                Report(actualExpression);
+                Report(actualOperation);
             }
         }
 
-        private ExpressionSyntax? GetExpectedExpression(InvocationExpressionSyntax assertExpression, IMethodSymbol methodSymbol, SemanticModel semanticModel)
+        private IOperation? GetExpectedExpression(IInvocationOperation assertOperation)
```
Comment (mikkelbu) [inline src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs:78]: Perhaps rename `GetExpectedExpression` to `GetExpectedOperation`?

Comment (Dreamescaper) [reply src/nunit.analyzers/ConstActualValueUsage/ConstActualValueUsageAnalyzer.cs]: Updated

**Comment by mikkelbu** [comment]
Thanks for the rewrite @Dreamescaper 👍, the performance improvements are quite significant.

**Pull request merged**